modules/dnn/src/dnn.cpp

```diff
@@ -4976,6 +4976,13 @@ void Net::setPreferableTarget(int targetId)
             bool fp16 = ocl::Device::getDefault().isExtensionSupported("cl_khr_fp16");
             if (!fp16 && targetId == DNN_TARGET_OPENCL_FP16)
                 impl->preferableTarget = DNN_TARGET_OPENCL;
+#endif
+        }
+        else if (IS_DNN_CUDA_TARGET(targetId))
+        {
+#ifdef HAVE_CUDA
+            if (!cuda4dnn::doesDeviceSupportFP16())
+                impl->preferableTarget = DNN_TARGET_CUDA;                
```

Comment:
I am unable to understand this fix. This check is already done in `initCUDABackend` and it also sends an error message. This check here will render that check useless.

Reply:
This will auto-correct target when setPreferableTarget(TARGET_CUDA_FP16) is called. A lot of things can happen after initCudaBackend, and there are no guarantees that the preferableTarget has not changed by the time `Net::forward()` is called. Or I'm missing something?

Reply:
@YashasSamaga Ok, I agree this does render your other check useless (and I can see that I don't have the auto-correct warning message when I add my fix). However it does indeed fixes some corner case because it won't complain about missing convolution for half anymore.

Reply:
Changing backend/target would reset the network which would force a reinitialization. I think there is a bug elsewhere and probably along the lines you're hinting at. Maybe something does indeed happen before the FP16->FP32 switch in initCUDABackend. This might fix the problem for now though but I think it's better to find the root cause than hide it.